ca_gazebo/src/color_sensor_plugin.cpp

```diff
@@ -0,0 +1,131 @@
+#include <gazebo/sensors/SensorTypes.hh>
+
+
+#include "std_msgs/Bool.h"
+
+
+namespace gazebo
+{
+  // Register this plugin with the simulator
+  GZ_REGISTER_SENSOR_PLUGIN(GazeboRosColor)
+
+  ////////////////////////////////////////////////////////////////////////////////
+  // Constructor
+  GazeboRosColor::GazeboRosColor():
+  _nh("color_sensor_plugin"),
+  _fov(6),
+  _range(10)
+{
+
+}
+  ////////////////////////////////////////////////////////////////////////////////
+  // Destructor
+  GazeboRosColor::~GazeboRosColor()
+  {
+    ROS_DEBUG_STREAM_NAMED("camera","Unloaded");
+  }
+
+  void GazeboRosColor::Load(sensors::SensorPtr _parent, sdf::ElementPtr _sdf)
+  {
+    // Make sure the ROS node for Gazebo has already been initialized
+    if (!ros::isInitialized())
+    {
+      ROS_FATAL_STREAM("A ROS node for Gazebo has not been initialized, unable to load plugin. "
+        << "Load the Gazebo system plugin 'libgazebo_ros_api_plugin.so' in the gazebo_ros package)");
+      return;
+    }
+
+    CameraPlugin::Load(_parent, _sdf);
+    // copying from CameraPlugin into GazeboRosCameraUtils
+    this->parentSensor_ = this->parentSensor;
+    this->width_ = this->width;
+    this->height_ = this->height;
+    this->depth_ = this->depth;
+    this->format_ = this->format;
+    this->camera_ = this->camera;
+    this->_rvalue_max = _sdf->GetElement("rvalue_max")->Get<double>();
+    this->_rvalue_min = _sdf->GetElement("rvalue_min")->Get<double>();
+    this->_gvalue_max = _sdf->GetElement("gvalue_max")->Get<double>();
+    this->_gvalue_min = _sdf->GetElement("gvalue_min")->Get<double>();
+    this->_bvalue_max = _sdf->GetElement("bvalue_max")->Get<double>();
+    this->_bvalue_min = _sdf->GetElement("bvalue_min")->Get<double>();
+    this->_publisher_name = _sdf->Get<std::string>("PublisherName");
+     _sensorPublisher = _nh.advertise<std_msgs::Bool>(this->_publisher_name, 1);
+     this->_factor_color_per_image=_sdf->GetElement("FactorColorperImage")->Get<double>();
+     this->parentSensor_->SetActive(true);
+
+    GazeboRosCameraUtils::Load(_parent, _sdf);
+  }
+
+  ////////////////////////////////////////////////////////////////////////////////
+  // Update the controller
+  void GazeboRosColor::OnNewFrame(const unsigned char *_image,
+    unsigned int _width, unsigned int _height, unsigned int _depth,
+    const std::string &_format)
+  {
+    static int seq=0;
+
+    int R,G,B;
+
+    double _desired_color_in_pixel=0, _pixel_count=0;
+
+    this->sensor_update_time_ = this->parentSensor_->LastUpdateTime();
+
+    if (!this->parentSensor->IsActive())
+    {
+      if ((*this->image_connect_count_) > 0)
+      // do this first so there's chance for sensor to run once after activated
+        this->parentSensor->SetActive(true);
+    }
+    else
+    {
+      if ((*this->image_connect_count_) > 0)
+      {
+        common::Time cur_time = this->world_->SimTime();
+        if (cur_time - this->last_update_time_ >= this->update_period_)
```

Comment:
This variable (`last_update_time_`) is never set before. You need to do it in `Load()`.